### ISSUE TITLE
Enable [or make a simple toggle for] local hostname resolution out of the box (mDNS, Avahi, llmnr, etc)

### ISSUE THREAD

**Issue by GabeAl** [clearlinux/distribution#2751]
One of the most common linux use-cases I've encountered is ssh-ing into machines on the network. Doing this usually requires their name, as the IP address changes in today's world (gone are the days of static IPs it seems). 

But this is broken out of the box, and despite many suggested workarounds, they again don't stick in the latest releases. Avahi doesn't work, setting LLMNR to "yes" doesn't work anymore, and my old workaround no longer persists after reboot. 

See: https://community.clearlinux.org/t/ssh-is-broken-by-default-how-to-fix-cant-connect-to-other-computers-by-hostname/7217

How do you guys use Clear Linux? Don't you ever need to ssh into another local machine? Do you remember the local IP addresses (e.g. 192.168.0.19) in your head for each one? Every other distro has this turned on (or an equivalent) out of the box so I can just do ssh@mylocalmachine2

This is so annoying, as after every reboot I have to gain a root shell and run
`for interface in $(ifconfig | grep "Ethernet" | sed 's/ .*//'); do systemd-resolve --set-mdns=yes --interface=$interface; done`

to get it to turn on. Then it works again until reboot. 

Can't we just get it on out of the box? Would be very helpful :)

**Comment by bryteise** [comment]
We view llmnr as a security risk that is a bad default for us to enable by default. If you want to change the distro default, https://www.freedesktop.org/software/systemd/man/resolved.conf.html describes how to configure overrides to distro content with /etc configuration files.

**Comment by thiagomacieira** [comment]
mDNS works fine for me. My Clear Linux can find other machines via it:

```
$ cat /etc/systemd/resolved.conf
[Resolve]
MulticastDNS=true
$ resolvectl query tjmaciei-mobl5.local   
tjmaciei-mobl5.local: <bunch of addresses go here>

-- Information acquired via protocol mDNS/IPv6 in 558us.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: cache
$ getent ahosts tjmaciei-mobl5.local
<bunch of addresses here>
$ ping -q -c4 tjmaciei-mobl5.local
PING tjmaciei-mobl5.local(tjmaciei-mobl5.local (<redacted>)) 56 data bytes

--- tjmaciei-mobl5.local ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3006ms
rtt min/avg/max/mdev = 2.115/2.317/2.667/0.209 ms
```

And it can be found via because of:
```
$ diff -u {/usr/lib,/etc}/systemd/network/80-dhcp.network                         
--- /usr/lib/systemd/network/80-dhcp.network    2020-05-15 23:12:42.000000000 +0000
+++ /etc/systemd/network/80-dhcp.network        2020-06-20 07:15:01.225500548 +0000
@@ -3,6 +3,7 @@
 
 [Network]
 DHCP=yes
+MulticastDNS=yes
 
 [DHCP]
 UseDomains=yes
```
(I don't know if systemd-networkd can do .network.d directories)

These are not the default, but they are not the default in upstream (systemd) either. It's very easy to turn it on for people who want this. But there are lot of people who won't want it, like a cloud instance deployment.

**Comment by GabeAl** [comment]
> 

No, if you look at the thread I linked, I do this and more, and it certainly does **not** persist across reboots at all. Because it does not work, I've come here. Because it does not work, pointing to an obscure system internals man page that ultimately says (to my understanding) "do what you've already tried" is less than a solution, but actually **evidence of the problem!**

Can you tell me what part of my attempted edits to the configuration were wrong/incomplete? 

> Think I solved it but can’t find this anywhere online:
> 
>     get a root shell (i.e. sudo su - )
>     Run this: printf "LLMNR=yes\nMulticastDNS=yes\n" >> /etc/systemd/resolved.conf; for interface in $(ifconfig | grep "Ethernet" | sed 's/ .*//'); do systemd-resolve --set-mdns=yes --interface=$interface; done
>     Reboot network: systemctl restart systemd-resolved
> 
It does not persist after reboot.

**Comment by thiagomacieira** [comment]
Again, there are two things you need to enable, because this is split between two systemd daemons:
* You need to configure the network interfaces you want mDNS to be used on (systemd-networkd)
* You need to configure the daemon that performs and responds to queries (systemd-resolved)

See the docs for [.network](https://www.freedesktop.org/software/systemd/man/systemd.network.html#MulticastDNS=) files and for [resolved's configuration](https://www.freedesktop.org/software/systemd/man/resolved.conf.html#MulticastDNS=). Both need to be configured. I get a feeling that resolved should default to true because everything is controlled per-interface by systemd.networkd, but the manual isn't clear what value is the default. And TBH, I don't understand why resolved needs this setting, if everything is controlled per-interface anyway.

There is an (AMI for Clear Linux on AWS)[https://aws.amazon.com/marketplace/pp/prodview-etajttzzjzs24], maintained by our team.

**Comment by GabeAl** [comment]
Thanks. 

Let's unpack my command above (the one I told you did not persist across reboots). 

First, I write: `printf "LLMNR=yes\nMulticastDNS=yes\n" >> /etc/systemd/resolved.conf` 
- This modifies the resolved.conf as above. 

Second, I write: `for interface in $(ifconfig | grep "Ethernet" | sed 's/ .*//'); do systemd-resolve --set-mdns=yes --interface=$interface; done`
- This sets each interface to mdns=yes.

The .network files don't work for me. They are too difficult to dive into and each one is dangerous to modify (if it exists at all), and there is no documentation on how to create them (at least none that work on a stateless system like Clear Linux that has nothing useful in these popular configuration paths). 

I want to revisit your statement earlier.  "It's very easy to turn it on for people who want this."

I still don't see an answer after a year, and your responses to read deep man pages does not make this "easy." Can you explain how it is done? Have you tried doing this yourself? How do you:
1. Create the network interface files
2. Determine which is correct (especially across users and systems! Don't tell me you expect someone to be intimately familiar with all computers!)
3. Edit the files and put the tag MulticastDNS=true in the right place that won't conflict with other settings in the file
4. Repeat this for every single interface on the computer
5. Update this and re-do it every time an interface is added/removed or simply updated

I'm sorry but I am going to have to correct you. No, this is far, far, far from easy. Do you understand now why I created this topic? Is it dawning on you why regular linux users who want to ssh into machines on their network cannot be expected to do this "easily"? In your view is it easy to compile and configure custom kernels? If so, I think your standards have drifted a bit, unfortunately. 

Make this easy for us. You want to ship an AMI for Amazon? Turn it off there. Everywhere else? Enable local name discovery. This is what people use. I don't buy it that "more people want it off" if every other major distro goes through pains to have it work out of the box. 

Please fix this, don't pretend it's "easy" and "everyone doesn't want it." It's not easy, and we do.

**Comment by bwarden** [comment]
`80-dhcp.network` is the default configuration for ethernet interfaces to use DHCP. You can tell because it defines a match for interfaces whose names start with "en". Check the `[Match]` section in all the other `.network` files in `/usr/lib/systemd/network` to see what I mean. Unless you've manually configured networking some other way, that's the file that defines your network interface configuration, so that's the file you have to override by copying it to `/etc/systemd/network/` with the exact same name and enabling `MulticastDNS`, just like you summarized in [your earlier comment](https://github.com/clearlinux/distribution/issues/2751#issuecomment-1270998487).

You _have to_ edit the `.network file`, and reload networking by `sudo networkctl reload`, or just reboot.

systemd is indeed complicated, and you really do have to read all the manpages to understand how and why it works, but @thiagomacieira distilled it to exactly what you need to do.
